bigmachine.go

```diff
@@ -362,13 +366,51 @@ func (b *B) pprofIndex(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (b *B) shutdownAllMachines(ctx context.Context, duration time.Duration) {
+	b.mu.Lock()
+	// Shutdown all of the existing machines.
+	for _, m := range b.machines {
+		// shutdown is best effort
+		err := m.Call(ctx, "Supervisor.Shutdown",
+			shutdownRequest{
+				Delay:   time.Second,
+				Message: string(logSyncMarker),
+			},
+			nil)
+		if err != nil {
+			log.Error.Printf("failed to invoke Supervisor.Shutdown on %v: %v\n",
+				m.Addr, err)
+		}
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(b.machines))
+	timeout := time.After(duration)
+	// Wait for the logs to propogate or for a timeout to occur.
+	for _, m := range b.machines {
+		go func(addr string, ch chan bool) {
+			select {
+			case tailed := <-ch:
+				if tailed {
+					log.Info.Printf("log closed for %v\n", addr)
+				}
+			case <-timeout:
+				log.Error.Printf("timeout after %v waiting for logs to close from: %v\n", duration, addr)
+			}
+			wg.Done()
+		}(m.Addr, m.tailDone)
+	}
```

Comment:
Suggestion:

```
ctx, cancel := .. context.WithTimeout(...)
g, ctx := errgroup.WithContext(ctx)
```

then use that and select on ctx.Done() as well as <-ch.

Reply:
thanks, I think it's a little cleaner with this.